docs/with-js.md

```diff
@@ -0,0 +1,173 @@
+
+```js
+import Component from '@glimmer/component';
+
+/**
+ * @typedef FancyInputArgs
+ * @property {'input' | 'textarea'} type The type of input to render
+ * @property {string} value The initial value to render into the input
+ * @property {(newValue: string) => void} onChange An action to run when the
+ *   input's value changes
+ */
+
+/**
+ * @typedef {HTMLInputElement | HTMLTextAreaElement} FancyInputElement
+ */
+
+/**
+ * @typedef FancyInputBlocks
+ * @property {[]} label
+ */
+
+/**
+ * @typedef FancyInputSignature
+ * @property {FancyInputArgs} Args
+ * @property {FancyInputElement} Element
+ * @property {FancyInputBlocks} Blocks
+ */
+
+/**
+ * A fancy `<input>` component, with styles pre-applied and some custom
+ * handling.
+ *
+ * @extends {Component<FancyInputSignature>}
+ */
+export default class FancyInput extends Component {
+  // implementation...
+}
+```
+
+Let's walk through this in detail. First, the `Args` to the component specify that it receives:
+
+- a `@type` argument specifying whether to render an `<input type='text'>` or a `<textarea>`
+- a `@value` argument, which should be a string
+- an `@onChange` action which receives the updated value and does something with it
+
+```js
+/**
+ * @typedef FancyInputArgs
+ * @property {'input' | 'textarea'} type The type of input to render
+ * @property {string} value The initial value to render into the input
+ * @property {(newValue: string) => void} onChange An action to run when the
+ *   input's value changes
+ */
+```
+
+The `@typedef` declaration for an `Element` must be a "union" type, declaring all the different types of element to which `...attributes` might be applied, with a pipe `|` as the "or" character for the different types. Here, we can imagine that the `...attributes` are applied to the input, and since it might be an `<input>` or a `<textarea>`, we write exactly that, using the names for the corresponding JS types:
+
+```js
+/**
+ * @typedef {HTMLInputElement | HTMLTextAreaElement} FancyInputElement
+ */
+```
+
+The component also provides a named block, `label`, which does not yield any value, allowing customization of the `<label>` for the `<input>`:
+
+```js
+/**
+ * @typedef FancyInputBlocks
+ * @property {[]} label
+ */
+```
+
+Finally, the component signature itself assembles those and also specifies the `Element`:
+
+```js
+/**
+ * A fancy `<input>` component, with styles pre-applied and some custom
+ * handling.
+ *
+ * @extends {Component<FancyInputSignature>}
+ */
+export default class FancyInput extends Component {
+  // implementation...
+}
+```
+
+Now this will provide useful completion, go-to-definition, and refactoring.
+
+### Caveats
+
+Undocumented component JS files will behave exactly as regular undocumented JS behaves: no information will be provided for editor support apart from what can be derived from the context.
```

Comment:
You definitely miss out on a lot by not writing doc comments, but I don't think we should undersell what value does exist here.

Type info, basic refactorings, and go-to-definition will work for anything hanging off of `this`, and inference based on things that _do_ have types (builtins, third-party dependencies with Glint types, etc) will work as a baseline without any extra typing.

Reply:
Good point. I'll add some text to that effect (this was just what was there in the original PR).

Reply:
```suggestion
Undocumented component JS files will behave exactly as regular undocumented JS behaves: no information will be provided for editor support apart from what can be derived from the context.

That caveat notwithstanding, Glint still offers quite a bit even without doc comments. Type info, basic refactorings, and go-to-definition will work for anything referring to `this`, and inference based on things that do have types (built-in items Glint knows about, third-party dependencies with Glint types, etc) will work as a baseline without any extra typing.
```